### ISSUE TITLE
Possible bug to convert string at the hash256

### ISSUE THREAD

**Issue by vincenzopalazzo** [nayuki/Bitcoin-Cryptography-Library#6]
Hello,

I to be convert a string "hello" in double hash256, I using your library and I converted the uint8_t[32] to string whit this code

```
std::string Sha256Hash::ToString()
{
  std::string hashResult;
  std::stringstream stream;
  for(int i = 0; i < HASH_LEN - 1; i++)
  {

    //stream << std::hex << static_cast<int>(value[i]); BUG
      int valueInt = static_cast<int>(value[i]);
      stream << std::hex << valueInt << std::setw(2) << std::setfill('0');

  }

  hashResult = stream.str();
  return hashResult;
}
```

The conversion is wrong, an example
 ```
string hello = "hello";

    Bytes byte = asciiBytes(hello.c_str());
    Sha256Hash shaHash = Sha256::getHash(byte.data(), byte.size());
    LOG(INFO) << "The string hello converting with sha256: " << shaHash.ToString();

    ASSERT_EQ(shaHash.ToString(), "2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824");
```
I get this result ```2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b98``` it'is wrong, because the last two characters are missing.

another example
```
string hello = "hello";
Bytes byte = asciiBytes(hello.c_str());
Sha256Hash shaHash = Sha256::getDoubleHash(byte.data(), byte.size());
LOG(INFO) << "The string hello converting with double sha256: " << shaHash.ToString();
ASSERT_EQ(shaHash.ToString(), "9595c9df9075148eb06860365df33584b75bff782a510c6cd4883a419833d50");
```
I get this result `9595c9df90075148eb06860365df33584b75bff782a510c6cd4883a419833d `it'is wrong, because there is an extra 0 in addition to the lack of the last two digits.

I may have wrong to string but I want to ask you if the error can be in the sha256 algorithm, I am completely ignorant in the cryptography field, if the algorithm can be bugged can you solve the problem?

I would like to use this library in a thesis work

*Thanks and sorry for my terrible English*

**Comment by nayuki** [comment]
You opened and closed this issue within a day, so I'll assume you solved the problem by yourself.

Nonetheless, I'm writing a late reply to provide a full explanation for future reference.

* My SHA-256 hasher implementation is correct and is covered by many test cases. I provided a constructor that converts a string to a `Sha256Hash` object, but I did not provide a method to convert that back to a string. You wrote your own `ToString()` method and showed the code in your opening message.

* You wrote:
    > it'is wrong, because the last two characters are missing

    This is because your `ToString()` function is wrong, because it counts `i` to `HASH_LEN - 1` (incorrect) instead of `HASH_LEN` (correct).

* You wrote:
    > it'is wrong, because there is an extra 0 in addition to the lack of the last two digits

    Your example is wrong; your `ASSERT_EQ()` has a 63-character hex string (incorrect) instead of 64 characters (the correct length for SHA-256 output).

In summary, you wrote a buggy `Sha256Hash::ToString()` method, wrote a buggy example using `getDoubleHash()`, then accused my library code of being wrong. After reviewing the code you provided, I find that the problem lies squarely in your code, not mine.

**Comment by vincenzopalazzo** [comment]
Hi @nayuki,

I apologize if my answer seemed bad or that you lack respect towards you, but it is absolutely not true, unfortunately my lack of knowledge of English made me look ugly.

Your library has allowed me to do various things, like:

1. Know a brilliant person, because only a brilliant person can write code like this.
2. It allowed me to create a project that works only with bitcoin blk files eliminating the rpc framework.
3. I became passionate about your every project, and I wanted to write to you because I imported this project into my thesis project and I wanted to share with you, for me it's great to be able to take the first steps on bitcoin.

I want to renew my apologies for my terrible English that perhaps made me look ugly in giving the question.

Ps: the errors were in my ToString () and not in your code, I had to use only a word other than wrong, sorry
